### PR TITLE
new-build command for creating BuildConfigs

### LINKED ISSUE #2695: new-build command that creates a new build
From the 1.0 Punch List #2628.

new-build command should be similar to new-app in that it should create *just* the `BuildConfig` object.

### PR DESCRIPTION
**Pull request by soltysh**
This should fix #2695.
@smarterclayton @csrwng ptal

This PR also includes a fix (2nd commit) for the case when generating new-app/new-build from repo containing Dockerfile (eg. https://github.com/openshift/ruby-hello-world) with forced build type (`--build=source`), previously that always created dockerStrategy.

### CODE REVIEW

File: pkg/cmd/cli/cmd/newapp.go [hunk 1]
```diff
@@ -160,10 +203,10 @@ func RunNewApplication(f *clientcmd.Factory, out io.Writer, c *cobra.Command, ar
 		}
 	}
 
-	if len(cmdutil.GetFlagString(c, "output")) != 0 {
-		return f.Factory.PrintObject(c, result.List, out)
-	}
+	return nil
+}
 
+func produceOutput(f *clientcmd.Factory, out io.Writer, result *newcmd.AppResult) error {
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newapp.go]: I think something like 'createObjects' would be a more appropriate name. produceOutput sounds like you're printing something.

File: pkg/cmd/cli/cmd/newapp.go [hunk 2]
```diff
@@ -132,20 +185,10 @@ func RunNewApplication(f *clientcmd.Factory, out io.Writer, c *cobra.Command, ar
...
-		if err == newcmd.ErrNoInputs {
-			// TODO: suggest things to the user
-			return cmdutil.UsageError(c, "You must specify one or more images, image streams, templates or source code locations to create an application.")
-		}
-		return err
-	}
+	return nil
+}
 
+func parseLabels(c *cobra.Command, result *newcmd.AppResult) error {
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newapp.go]: applyLabels or setLabels

File: pkg/cmd/cli/cmd/newapp.go [hunk 3]
```diff
@@ -106,6 +106,59 @@ func NewCmdNewApplication(fullName string, f *clientcmd.Factory, out io.Writer)
...
+				}
+				hasMissingRepo = true
+				fmt.Fprintf(c.Out(), "WARNING: We created an ImageStream %q, but it does not look like a Docker registry has been integrated with the OpenShift server. Automatic builds and deployments depend on that integration to detect new images and will not function properly.\n", t.Name)
+			}
+		}
+	}
+	return nil
+}
+
+func parseParameters(f *clientcmd.Factory, c *cobra.Command, args []string, config *newcmd.AppConfig) error {
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newapp.go]: setupAppConfig

File: pkg/cmd/cli/cmd/newapp.go [hunk 4]
```diff
@@ -106,6 +106,59 @@ func NewCmdNewApplication(fullName string, f *clientcmd.Factory, out io.Writer)
...
+	}
+
+	result, err := config.RunAll(out)
+	if err != nil {
+		if errs, ok := err.(errors.Aggregate); ok {
+			if len(errs.Errors()) == 1 {
+				err = errs.Errors()[0]
+			}
+		}
+		if err == newcmd.ErrNoInputs {
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newapp.go]: this should be after the enclosing if

File: pkg/cmd/cli/cmd/newbuild.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/errors"
+	"github.com/spf13/cobra"
+
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	newcmd "github.com/openshift/origin/pkg/generate/app/cmd"
+)
+
+const (
+	newBuildLong = `Create a new build in OpenShift by specifying source code, and/or images.
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: ...source code, and/or images. -> source code.

(it's either source code or source code + image)

File: pkg/cmd/cli/cmd/newbuild.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+This command will try to create a BuildConfig for your application using images, or code that has
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: ...using images, or code -> ...using images and code

File: pkg/cmd/cli/cmd/newbuild.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+a public repository located on your system. It will lookup the images on the
+local Docker installation (if available), a Docker registry, or an OpenShift image stream.
+If you specify a source code URL, it will set up a build that takes your source code and converts
+it into an image that can run inside of a pod. Local source must be in a git repository that has
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: s/has/has a/

File: pkg/cmd/cli/cmd/newbuild.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	newcmd "github.com/openshift/origin/pkg/generate/app/cmd"
+)
+
+const (
+	newBuildLong = `Create a new build in OpenShift by specifying source code, and/or images.
+
+This command will try to create a BuildConfig for your application using images, or code that has
+a public repository located on your system. It will lookup the images on the
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: not necessarily located on your system... you can give it a URL

File: pkg/cmd/cli/cmd/newbuild.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+local Docker installation (if available), a Docker registry, or an OpenShift image stream.
+If you specify a source code URL, it will set up a build that takes your source code and converts
+it into an image that can run inside of a pod. Local source must be in a git repository that has
+remote repository that the OpenShift instance can see.
+
+Once the BuildConfig will be created you may need to run a build with 'start-build'.`
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: BuildConfig will be -> BuildConfig is

File: pkg/cmd/cli/cmd/newbuild.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+	newBuildExample = `  // Create a BuildConfig based on the source code in the current git repository (with a public remote) and a Docker image
+  $ %[1]s new-build . --docker-image=repo/langimage
+
+  // Create a NodeJS application based on the provided [image]~[source code] combination
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: s/application/build configuration/

File: pkg/cmd/cli/cmd/newbuild.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+This command will try to create a BuildConfig for your application using images, or code that has
+a public repository located on your system. It will lookup the images on the
+local Docker installation (if available), a Docker registry, or an OpenShift image stream.
+If you specify a source code URL, it will set up a build that takes your source code and converts
+it into an image that can run inside of a pod. Local source must be in a git repository that has
+remote repository that the OpenShift instance can see.
+
+Once the BuildConfig will be created you may need to run a build with 'start-build'.`
+
+	newBuildExample = `  // Create a BuildConfig based on the source code in the current git repository (with a public remote) and a Docker image
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: s/BuildConfig/build configuration/

File: pkg/cmd/cli/cmd/newbuild.go [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+  $ %[1]s new-build . --docker-image=repo/langimage
+
+  // Create a NodeJS application based on the provided [image]~[source code] combination
+  $ %[1]s new-build openshift/nodejs-010-centos7~https://bitbucket.com/user/nodejs-app
+
+  // Create an application from a remote repository using its beta2 branch
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: s/application/build configuration/

File: pkg/cmd/cli/cmd/newapp.go [hunk 13]
```diff
@@ -32,12 +32,12 @@ const (
 	newAppLong = `Create a new application in OpenShift by specifying source code, templates, and/or images.
 
 This command will try to build up the components of an application using images, templates,
-or code that has a public repository located on your system. It will lookup the images on the 
-local Docker installation (if available), a Docker registry, or an OpenShift image stream. 
-If you specify a source code URL, it will set up a build that takes your source code and converts 
-it into an image that can run inside of a pod. Local source must be in a git repository that has 
-remote repository that the OpenShift instance can see. The images will be deployed via a deployment 
-configuration, and a service will be hooked up to the first public port of the app. You may either specify 
+or code that has a public repository located on your system. It will lookup the images on the
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newapp.go]: I would remove 'located on your system.'

File: pkg/cmd/cli/cmd/newapp.go [hunk 14]
```diff
@@ -32,12 +32,12 @@ const (
...
+local Docker installation (if available), a Docker registry, or an OpenShift image stream.
+If you specify a source code URL, it will set up a build that takes your source code and converts
+it into an image that can run inside of a pod. Local source must be in a git repository that has
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newapp.go]: s/has/has a/

File: pkg/cmd/cli/cmd/newapp.go [hunk 15]
```diff
@@ -32,12 +32,12 @@ const (
...
+remote repository that the OpenShift instance can see. The images will be deployed via a deployment
+configuration, and a service will be hooked up to the first public port of the app. You may either specify
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newapp.go]: s/hooked up/connected/

File: pkg/cmd/cli/cmd/newbuild.go [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+  $ %[1]s new-build https://github.com/openshift/ruby-hello-world#beta2`
+)
+
+// NewCmdNewBuild implements the OpenShift cli new-build command
+func NewCmdNewBuild(fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	_, typer := f.Object()
+	config := newcmd.NewAppConfig(typer)
+
+	cmd := &cobra.Command{
+		Use:     "new-build (IMAGE | IMAGESTREAM | TEMPLATE | PATH | URL ...)",
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: TEMPLATE should not be in the list

File: pkg/cmd/cli/cmd/newbuild.go [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+		Short:   "Create a new BuildConfig",
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: s/BuildConfig/build configuration/

File: pkg/cmd/cli/cmd/newbuild.go [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+		Run: func(c *cobra.Command, args []string) {
+			err := RunNewBuild(f, out, c, args, config)
+			if err == errExit {
+				os.Exit(1)
+			}
+			cmdutil.CheckErr(err)
+		},
+	}
+
+	cmd.Flags().Var(&config.SourceRepositories, "code", "Source code to use to build this application.")
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: "to build this application" -> "in the build configuration"

File: pkg/cmd/cli/cmd/newbuild.go [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+	cmd.Flags().VarP(&config.ImageStreams, "image", "i", "Name of an OpenShift image stream to use in the app.")
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: "to use in the app" -> "to use as a builder"

File: pkg/cmd/cli/cmd/newbuild.go [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+	cmd.Flags().Var(&config.DockerImages, "docker-image", "Name of a Docker image to include in the app.")
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: "to include in the app" -> "to use as a builder"

File: pkg/cmd/cli/cmd/newbuild.go [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+	cmd.Flags().StringVar(&config.Name, "name", "", "Set name to use for generated application artifacts")
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: s/application artifacts/build artifacts"

File: pkg/cmd/cli/cmd/newbuild.go [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+	cmd.Flags().StringVar(&config.TypeOfBuild, "build", "", "Specify the type of build to use if you don't want to detect (docker|source).")
+	cmd.Flags().StringP("labels", "l", "", "Label to set in all resources for this application.")
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: ... to set in all generated resources.

File: pkg/cmd/cli/cmd/newbuild.go [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+
+	cmd.Flags().StringP("output", "o", "", "Output format. One of: json|yaml.") //|template|templatefile.")
+	cmd.Flags().String("output-version", "", "Output the formatted object with the given version (default api-version).")
+	cmd.Flags().Bool("no-headers", false, "When using the default output, don't print headers.")
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: Remove

File: pkg/cmd/cli/cmd/newbuild.go [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+	cmd.Flags().String("template", "", "Template string or path to template file to use when -o=template or -o=templatefile.  The template format is golang templates [http://golang.org/pkg/text/template/#pkg-overview]")
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: Remove - Unless we want to support templates in new-build

File: pkg/cmd/cli/cmd/newbuild.go [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+	result, err := config.RunBuilds(out)
+	if err != nil {
+		if errs, ok := err.(errors.Aggregate); ok {
+			if len(errs.Errors()) == 1 {
+				err = errs.Errors()[0]
+			}
+		}
+		if err == newcmd.ErrNoInputs {
+			// TODO: suggest things to the user
+			return cmdutil.UsageError(c, "You must specify one or more images, image streams or source code locations to create a BuildConfig.")
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: or source code locations -> and source code locations

File: pkg/generate/app/cmd/newapp.go [hunk 26]
```diff
@@ -418,10 +418,21 @@ type AppResult struct {
...
+	return c.run(out, app.Acceptors{app.NewAcceptUnique(c.typer), app.AcceptNew})
+}
+
+// RunBuilds executes the provided config to generate just builds.
+func (c *AppConfig) RunBuilds(out io.Writer) (*AppResult, error) {
+	return c.run(out, app.Acceptors{app.NewAcceptBuildConfigs(c.typer), app.NewAcceptUnique(c.typer), app.AcceptNew})
+}
+
+// run executes the provided config applying provided acceptors.
+func (c *AppConfig) run(out io.Writer, acceptors app.Acceptors) (*AppResult, error) {
```
Comment (csrwng) [inline pkg/generate/app/cmd/newapp.go]: Line 444 - if you asked for NewBuild and you don't have source you should return an error.

File: pkg/generate/app/pipeline.go [hunk 27]
```diff
@@ -311,6 +311,30 @@ func objectMetaData(raw interface{}) (runtime.Object, *kapi.ObjectMeta, error) {
...
+func (a *acceptBuildConfigs) Accept(from interface{}) bool {
+	obj, _, err := objectMetaData(from)
+	if err != nil {
+		return false
+	}
+	_, kind, err := a.typer.ObjectVersionAndKind(obj)
+	if err != nil {
+		return false
+	}
+	return kind == "BuildConfig"
```
Comment (csrwng) [inline pkg/generate/app/pipeline.go]: What about the output image stream? or the input image stream if creating from a docker repo?

**Comment by csrwng** [comment]
I don't see a need for new-build to handle templates, since you can already do that with new-app, plus only creating buildConfigs from a template might not make sense. If that's the case, you need to add code in new-app.go to not allow templates when running in new-build mode.

**Comment by smarterclayton** [comment]
Agreed

----- Original Message -----

File: pkg/generate/app/pipeline.go (hunk 27 above)
Comment (smarterclayton) [inline pkg/generate/app/pipeline.go]: Always end an output image stream from new-build, unless we support an argument that lets you push to an external Docker repo (nice to have, can be post 3.0)

----- Original Message -----

File: pkg/cmd/cli/cmd/newbuild.go (hunk 16 above)
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: IMAGE and IMAGESTREAM should be there only if associated with SOURCE

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/newbuild.go]: It's ok to have "IMAGE SOURCE" - the normal logic should associate them.

----- Original Message -----

File: pkg/cmd/cli/cmd/newbuild.go (hunk 9 above)
Comment (csrwng) [inline pkg/cmd/cli/cmd/newbuild.go]: s/BuildConfig/build configuration/

File: pkg/cmd/cli/cmd/newapp.go (hunk 4 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/newapp.go]: What's wrong with this being here, at this point you are certain that the `err != nil` and you're just checking its type?

File: pkg/cmd/cli/cmd/newbuild.go (hunk 9 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/newbuild.go]: Throughout all the logs/msgs we keep original object name, thus here I'll leave BuildConfig.

Comment (soltysh) [inline pkg/cmd/cli/cmd/newbuild.go]: It's easier to find in docs BuildConfig than build configuration.

File: pkg/cmd/cli/cmd/newbuild.go (hunk 11 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/newbuild.go]: Same as above, leaving BuildConfig.

File: pkg/cmd/cli/cmd/newbuild.go (hunk 12 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/newbuild.go]: Will be BuildConfig.

File: pkg/cmd/cli/cmd/newapp.go (hunk 13 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/newapp.go]: Done in both newapp.go and newbuild.go

File: pkg/cmd/cli/cmd/newbuild.go (hunk 17 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/newbuild.go]: Same as above, leaving BuildConfig for clarity.

File: pkg/cmd/cli/cmd/newbuild.go (hunk 24 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/newbuild.go]: I'd leave it, without it you cannot use -o json|yaml which is useful if you just want to see the objects created. The printer internally reaches for template, without it you'll get an error that you're trying to access flag that wasn't defined, been there and tested that.

File: pkg/generate/app/pipeline.go (hunk 27 above)
Comment (soltysh) [inline pkg/generate/app/pipeline.go]: We do support pushes to external repo -> DockerImage.

**Comment by soltysh** [comment]
If we remove templates, I need to remove --output as well, see [this comment](https://github.com/openshift/origin/pull/2747#discussion_r31675254)

File: pkg/cmd/cli/cmd/newbuild.go (hunk 24 above)
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/newbuild.go]: We don't want to support real templates - you can go back to using the AddPrinterFlags function

**Comment by soltysh** [comment]
@csrwng addressed most of your comments for others I wrote my reasoning. I've also fixed the tests that was failing. The only left question is whether we want `--output` and `--templates` flags.

File: pkg/generate/app/pipeline.go (hunk 27 above)
Comment (smarterclayton) [inline pkg/generate/app/pipeline.go]: It's probably not going to be the default - a user should have to give us a --to-docker flag (or similar).  Ideally I'd like to reserve "--to" to be an Openshift image tag (so you can easily spin up a new build pointing to that tag, vs using a default name).

File: pkg/cmd/cli/cmd/newapp.go [hunk 28]
```diff
@@ -106,6 +106,59 @@ func NewCmdNewApplication(fullName string, f *clientcmd.Factory, out io.Writer)
...
+	result, err := config.RunAll(out)
+	if err != nil {
+		if errs, ok := err.(errors.Aggregate); ok {
+			if len(errs.Errors()) == 1 {
+				err = errs.Errors()[0]
+			}
+		}
+		if err == newcmd.ErrNoInputs {
+			// TODO: suggest things to the user
+			return cmdutil.UsageError(c, "You must specify one or more images, image streams, templates or source code locations to create an application.")
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/newapp.go]: Remove templates

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/newapp.go]: Oh, this is the new code

Comment (soltysh) [inline pkg/cmd/cli/cmd/newapp.go]: Yes, this is part of newapp.go, so this should be as is.

File: pkg/generate/app/pipeline.go (hunk 27 above)
Comment (soltysh) [inline pkg/generate/app/pipeline.go]: I'll catch you tomorrow on IRC to talk this through.

**Comment by soltysh** [comment]
@smarterclayton I've changed newbuild.go to use `AddPrinterFlags`. The only thing left will be then `--to-docker` flag I'm hoping to discuss with you tomorrow.

File: pkg/cmd/cli/cmd/newapp.go [hunk 29]
```diff
@@ -106,6 +106,59 @@ func NewCmdNewApplication(fullName string, f *clientcmd.Factory, out io.Writer)
...
+	for _, item := range result.List.Items {
+		switch t := item.(type) {
+		case *kapi.Service:
+			portMappings := "."
+			if len(t.Spec.Ports) > 0 {
+				portMappings = fmt.Sprintf(" with port mappings %s.", describeServicePorts(t.Spec))
+			}
+			fmt.Fprintf(c.Out(), "Service %q created at %s%s\n", t.Name, t.Spec.PortalIP, portMappings)
+		case *buildapi.BuildConfig:
+			fmt.Fprintf(c.Out(), "A build was created - you can run `osc start-build %s` to start it.\n", t.Name)
```
Comment (mfojtik) [inline pkg/cmd/cli/cmd/newapp.go]: don't we have constant for `osc` string?

File: pkg/cmd/cli/cmd/newbuild.go [hunk 30]
```diff
@@ -0,0 +1,107 @@
...
+			if err == errExit {
+				os.Exit(1)
+			}
+			cmdutil.CheckErr(err)
+		},
+	}
+
+	cmd.Flags().Var(&config.SourceRepositories, "code", "Source code in the build configuration.")
+	cmd.Flags().VarP(&config.ImageStreams, "image", "i", "Name of an OpenShift image stream to to use as a builder.")
+	cmd.Flags().Var(&config.DockerImages, "docker-image", "Name of a Docker image to use as a builder.")
```
Comment (mfojtik) [inline pkg/cmd/cli/cmd/newbuild.go]: DockerImages vs. 'docker-image'

File: pkg/cmd/cli/cmd/newbuild.go [hunk 31]
```diff
@@ -0,0 +1,107 @@
...
+	cmd.Flags().StringVar(&config.Name, "name", "", "Set name to use for generated build artifacts")
+	cmd.Flags().StringVar(&config.TypeOfBuild, "build", "", "Specify the type of build to use if you don't want to detect (docker|source).")
```
Comment (mfojtik) [inline pkg/cmd/cli/cmd/newbuild.go]: build-type ?

File: pkg/cmd/cli/cmd/newapp.go (hunk 4 above)
Comment (csrwng) [inline pkg/cmd/cli/cmd/newapp.go]: sorry, yup makes sense.... for some reason I thought it was inside the previous if

**Comment by smarterclayton** [comment]
We want --output and --templates like osc new-app yes - all our "generate" commands should be able to spit out yaml so that someone can play around and try stuff.

----- Original Message -----

File: pkg/cmd/cli/cmd/newapp.go (hunk 29 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/newapp.go]: Fixed.

File: pkg/cmd/cli/cmd/newbuild.go (hunk 30 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/newbuild.go]: Internally DockerImages allows having more than one.

File: pkg/cmd/cli/cmd/newbuild.go (hunk 31 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/newbuild.go]: Agreed on IRC to rename that to strategy, to match the produced output.

**Comment by soltysh** [comment]
@smarterclayton added `--to-docker` flag.
@csrwng I've renamed `--build` to `--strategy` as discussed on IRC.
ptal
[test] in the meantime

File: pkg/cmd/cli/cmd/newbuild.go [hunk 32]
```diff
@@ -0,0 +1,109 @@
...
+
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	newcmd "github.com/openshift/origin/pkg/generate/app/cmd"
+)
+
+const (
+	newBuildLong = `Create a new build in OpenShift by specifying source code.
+
+This command will try to create a BuildConfig for your application using images and
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/newbuild.go]: `build configuration (BuildConfig)`

Comment (soltysh) [inline pkg/cmd/cli/cmd/newbuild.go]: Fixed.

**Comment by soltysh** [comment]
Updated and fixed test error.

**Comment by soltysh** [comment]
I don't get the error, it's not related to this PR.
re-[test]

**Comment by soltysh** [comment]
Again, rebased and waiting for tests result.

**Comment by smarterclayton** [comment]
LGTM [merge]

**Pull request merged**